src/v/redpanda/application.cc

```diff
@@ -264,7 +264,11 @@ static void set_auditing_kafka_client_defaults(
 application::application(ss::sstring logger_name)
   : _log(std::move(logger_name)) {};
 
-application::~application() = default;
+application::~application() {
+    while (!_deferred.empty()) {
+        _deferred.pop_back();
```

Comment:
what is in this list that is added prior to app.run starting the first root fiber?

Reply:
not sure I follow the question here fully, the bug here is that  we are relying on the order of member definitions in Application class at shutdown (particularly in cases where there is no orderly shutdown, exception etc), that seems bug prone.

There are sharded members defined after _deferred, for example..

```
 deferred_actions _deferred;

....

ss::sharded<ss::abort_source> _as;
```

Without an orderly shutdown, when the default destructor is called, _as is destroyed first (without stopping it) and that triggers an assert, rather the fix attempts to explicitly invoke deferred actions, if any and then do the default cleanup.

Another minor issue is the destructor of `deque` that backs deferred_actions destroys the elements from first to last whereas we prefer the opposite order due to service dependencies.

Reply:
in the application destructor, seastar may have already been shutdown, right? seems like any cleanup should happen in the try/catch block in app:run?

Reply:
thats a good point, I thought the ss lifetime is tied to the Application class but looks like its a local variable inside run(), makes me wonder why we don't have even weirder stack traces on crashes (I'd expect a seg fault on accessing some reactor state).